zircon.core/src/main/kotlin/org/hexworks/zircon/internal/component/impl/log/LogElement.kt

```diff
@@ -0,0 +1,128 @@
+        val logElementRenderInfo = mutableListOf<Pair<Position, Int>>()
+        var delayInMs = context.delayTimeInMs
+
+        getWordsOfLogElement()
+                .forEach { word ->
+                    if (getPosition().y > currentLogElementY)
+                        currentScreenPosY += getPosition().y - currentLogElementY
+                    if (logArea.wrapLogElements && (currentScreenPosX + word.length) > tileGraphics.size.width()) {
+                        currentScreenPosX = 0
+                        currentScreenPosY += 1
+                    }
+
+                    val position = Position.create(currentScreenPosX, currentScreenPosY)
+                    logElementRenderInfo.add(Pair(position, word.length))
+                    if (visibleRenderArea.contains(position)) {
+                        val wordContext = LogElementRenderContext(context.componentRenderContext, context.positionRenderContext, delayInMs)
+                        renderWord(position, wordContext, tileGraphics, word)
+                        delayInMs += word.length * logArea.delayInMsForTypewriterEffect
+                        currentScreenPosX += word.length
+                        currentLogElementY = getPosition().y
+                    }
+                }
+
+        if (logElementRenderInfo.isNotEmpty()) {
+            val startRenderPosition = logElementRenderInfo.first().first
+            val endRenderPosition = logElementRenderInfo.last().first.plus(Position.create(logElementRenderInfo.last().second, 0))
+            screenPositionArea = ScreenPositionArea(startRenderPosition, endRenderPosition)
+        }
+
+        return LogElementRenderResult(currentScreenPosY, delayInMs)
+    }
+
+    private fun renderWord(position: Position, context: LogElementRenderContext, tileGraphics: SubTileGraphics,
+                           word: String) {
+        val visiblePosition = position.minus(context.componentRenderContext.component.visibleOffset)
+        var delayTimeInMs = context.delayTimeInMs
+        word.forEachIndexed { col, char ->
+            val wordModifiers = if (delayTimeInMs == 0L)
+                modifiers
+            else
+                modifiers.plus(Delay(delayTimeInMs))
+            tileGraphics.setTileAt(visiblePosition.withRelativeX(col), TileBuilder
+                    .newBuilder()
+                    .character(char)
+                    .styleSet(context.componentRenderContext.componentStyle.currentStyle())
+                    .modifiers(wordModifiers)
+                    .build())
+            delayTimeInMs += context.componentRenderContext.component.delayInMsForTypewriterEffect
+        }
+    }
+
+    private fun getWordsOfLogElement(): List<String> {
+        return if (text != "")
+            text.split(" ").map { "$it " }
+        else
+            listOf("")
+    }
+
+}
+
+data class LogComponentElement(val component: Component, private val xPosition: Int) : LogElement(xPosition) {
```

Comment:
This can be a combination of `Label`s and the `Component`. It would simplify the code a bit. I'd also add text + `Component` to this one as a single `Component`. This means that I'd add the `Label` + `Button` + `Label` to a `Panel` and add the `Panel` itself to the `LogArea`. This can greatly simplify how positions are calculated.